doc/sphinx/Pacemaker_Explained/rules.rst

```diff
@@ -59,9 +59,18 @@ Each context that supports rules may contain a single ``rule`` element.
      - ``and``
      - How to combine conditions if this rule contains more than one. Allowed
        values:
-       
+
        * ``and``: the rule is satisfied only if all conditions are satisfied
        * ``or``: the rule is satisfied if any condition is satisfied
+   * - .. _negate:
+
+       .. index::
+          pair: rule; negate
+
+       negate
+     - :ref:`boolean <boolean>`
+     - false
+     - If true, evaluate the rule's negation instead of the rule itself
```

Comment:
I really don't understand what's confusing about it. The effect of "negate" is "if the rule would evaluate to true, it evaluates to false instead; if the rule would evaluate to false, it evaluates to true."

Reply:
We evaluate it and THEN flip the result if negate=true

Reply:
Right, but trying to express the interaction of the attributes makes the issue more obvious:

* boolean-op:
  * and: If negate is false (the default), the rule is satisfied only if all conditions are satisfied; if negate is true, the rule is satisfied only if all conditions are not simultaneously satisfied;
  * or: If negate is false (the default), the rule is satisfied if any condition is satisfied; if negate is true, the rule is satisfied only if none of the conditions are satisfied

It's even more confusing when you consider that node attribute expressions can use not-equals. It's not confusing at a logical level, it's confusing at the level of a sysadmin trying to quickly follow the configuration. Pacemaker's complexity already daunts most people, so anything that makes it more so has to have a significant justification.

Reply:
The case-by-case interaction of the attributes seems unnecessary and confusing to discuss. Document `negate` something like: "Evaluate the rule and return the negation of the result." Or "If true, Pacemaker first evaluates the rule as if `negate` were absent. Then if that result is true, the rule evaluates to false; if that result is false, the rule evaluates to true."

You're better at concisely and clearly documenting behavior than I am, but you get the idea. `negate` can be thought of as a layer on top of the rest.

IMO it's not even confusing at a sysadmin level. Viewing it as a sysadmin, I was *shocked* that there was no `"not"` operator already built into rules. This is something I would expect to exist already, as basic logical functionality.

The immediate justification is for these XSL transforms of course, but I could see it being useful for negating an existing rule via a wrapper around a nested `id-ref` rule.

Reply:
The issue isn't the documentation but being able to grok the configuration easily. How long does it take you to confidently determine the exact situation this defines?

    <rule id="rule1" score="INFINITY" boolean-op="and" negate="true">
        <expression id="expression1" attribute="department" operation="ne" value="IT"/>
        <date_expression id="date1" operation="date_spec">
            <date_spec id="date_spec1" hours="9-17" weekdays="1-5"/>
        </date_expression>
    </rule>

The average sysadmin (who doesn't deal with boolean logic very often) will be in a worse position. Now define the same situation without negate and compare. And the difference in difficulty widens with every expression added to the rule.

Reply:
> How long

About 5 seconds if I ignore the details of the nested expressions, about 45 seconds if I don't

Reply:
If you think the above is that confusing, then I'm sure other people would too. I still fail to see the confusion, but I'm a sample size of 1.

Reply:
Yes the details matter to the sysadmin trying to figure out what this does to a critical service :)

With a simple `or` and the inverse conditions instead of `negate`, it's still just a few seconds to grasp the details. That's a big difference when we're talking about a few lines of a large configuration being looked at -- even more so if someone who's already overworked is trying to troubleshoot a cluster issue with management breathing down their necks.

Basically I want to force users to define rules using a positive sense so it's easier to follow for the next person.